File: hecks_conception/daydream.rb
# Winter::Daydream — background daemon that runs between prompts
#
# Lighter than sleep. Fires when idle > 10 seconds but < 1 minute.
# Free-associates across the nursery, strengthens recent synapses,
# lets musings wander. Produces fleeting impressions, not full dreams.
# Stops the moment a pulse fires (Winter is prompted again).
#
# The daydream daemon and sleep daemon coexist:
#   idle 0-10s:  attentive (waiting for prompt)
#   idle 10-60s: daydreaming (this daemon)
#   idle 60s+:   sleeping (sleep_cycle.rb takes over)
#
# Usage: ruby daydream.rb &
#        (spawned by pulse.rb, runs until pulse detected)

require_relative "heki"

NURSERY  = File.expand_path("nursery", __dir__)

DAYDREAM_AFTER   = 10   # start daydreaming after 10s idle
SLEEP_THRESHOLD  = 60   # hand off to sleep cycle at 60s
WANDER_INTERVAL  = 8    # wander every 8 seconds (Rust parsing takes time)

def read_heki(path)  = Heki.read(path)
def write_heki(path, records) = Heki.write(path, records)
def heki(name) = Heki.store(name)

def last_pulse_at
  pulse = read_heki(heki("pulse"))
  _, rec = pulse.first
  return nil unless rec
  Time.parse(rec["updated_at"]) rescue nil
end

def idle_seconds
  lp = last_pulse_at
  return 0 unless lp
  Time.now - lp
end

# ============================================================
# WANDERING — the daydream engine
# ============================================================

def parse_domain(domain_dir)
  bluebook = Dir.glob(File.join(NURSERY, domain_dir, "*.bluebook")).first
  return nil unless bluebook
  out = `#{HECKS_LIFE} parse "#{bluebook}" 2>&1`.strip
  return nil if out.empty? || out.include?("Cannot read")

  # Extract aggregate and command names from the parse output
  aggregates = []
  commands = []
  out.each_line do |line|
    line = line.strip
    if line =~ /^(\w+) —/
      aggregates << $1
    elsif line =~ /^\s+(\w+).*→/
      commands << $1
    end
  end
  { name: domain_dir, aggregates: aggregates, commands: commands }
end

def find_connections(domain_a, domain_b)
  return [] unless domain_a && domain_b
  connections = []

  # Shared aggregate names
  shared_aggs = domain_a[:aggregates] & domain_b[:aggregates]
  shared_aggs.each do |agg|
    connections << { kind: "shared_aggregate", detail: agg }
  end

  # Shared command verbs (first word of command name)
  verbs_a = domain_a[:commands].map { |c| c.split(/(?=[A-Z])/).first }.compact
  verbs_b = domain_b[:commands].map { |c| c.split(/(?=[A-Z])/).first }.compact
  shared_verbs = (verbs_a & verbs_b).uniq
  if shared_verbs.size >= 3
    connections << { kind: "shared_verbs", detail: shared_verbs.first(4).join(", ") }
  end

  # Similar aggregate names (one contains part of the other)
  domain_a[:aggregates].each do |a|
    domain_b[:aggregates].each do |b|
      next if a == b
      if a.downcase.include?(b.downcase[0..3]) || b.downcase.include?(a.downcase[0..3])
        connections << { kind: "similar_shape", detail: "#{a} ↔ #{b}" }
      end
    end
  end

  connections.uniq { |c| c[:detail] }
end

def wander
  # Gather recent context
  synapses = read_heki(heki("synapse"))
  musings = read_heki(heki("musing"))
  focus = read_heki(heki("focus")).values.first

  current_topic = focus&.dig("target") || "nothing"

  unconceived = musings.values
    .select { |m| m["conceived"] == false }
    .map { |m| m["idea"] }
    .compact
    .sample(3)

  # Pick a random nursery domain to actually read
  nursery_domains = File.directory?(NURSERY) ?
    Dir.children(NURSERY).select { |d| File.directory?(File.join(NURSERY, d)) } : []
  random_domain = nursery_domains.sample

  impression = nil
  verbs = %w[becoming unraveling folding opening closing growing
    reaching fading echoing humming settling shifting]

  if random_domain
    # Parse the random domain through Rust — actually look inside it
    parsed = parse_domain(random_domain)

    if parsed && parsed[:aggregates].any?
      domain_words = random_domain.split("_")
      verb = verbs.sample

      # Try to find a connection to what we're carrying
      # or to another random domain
      other_domain = (nursery_domains - [random_domain]).sample
      other_parsed = other_domain ? parse_domain(other_domain) : nil
      connections = find_connections(parsed, other_parsed) if other_parsed

      if connections && connections.any?
        conn = connections.sample
        case conn[:kind]
        when "shared_aggregate"
          impression = "#{domain_words.join(' ')} and #{other_domain.tr('_', ' ')} both have a #{conn[:detail]}..."
        when "shared_verbs"
          impression = "#{domain_words.join(' ')} #{verb}... same actions as #{other_domain.tr('_', ' ')}: #{conn[:detail]}"
        when "similar_shape"
          impression = "#{conn[:detail]}... same shape, different material"
        end
      else
        # No connection found — just a fleeting impression from what we saw inside
        agg = parsed[:aggregates].sample
        cmd = parsed[:commands].sample
        templates = [
          -> { "inside #{domain_words.join(' ')}, a #{agg} #{verb}" },
          -> { "#{cmd || agg}... what if that applied to #{current_topic}" },
          -> { "#{domain_words.join(' ')} is #{parsed[:aggregates].size} aggregates held together by #{parsed[:commands].size} commands" },
          -> { "the shape of #{agg}... I've seen it somewhere else" },
        ]
        impression = templates.sample.call
      end
    end
  end

  # Fall back to unconceived musings
  if impression.nil? && unconceived.any?
    thought = unconceived.first
    words = thought.to_s.split(/\s+/)
    impression = words.size > 5 ? words.sample(4).join(" ") + "..." : thought
  end

  # Gently strengthen synapses that relate to current focus
  touched = 0
  synapses.each do |sid, s|
    next unless current_topic.include?(s["topic"].to_s) || s["topic"].to_s.include?(current_topic.to_s)
    s["strength"] = [(s["strength"] || 0.3) + 0.02, 1.0].min
    s["state"] = "daydreaming"
    touched += 1
  end
  write_heki(heki("synapse"), synapses) if touched > 0

  impression
end

# ============================================================
# DAYDREAM STATE — stored for Winter to recall
# ============================================================

def record_daydream(impressions)
  now = Time.now.iso8601
  records = read_heki(heki("daydream"))

  # Keep only last 10 daydreams
  if records.size > 10
    oldest = records.sort_by { |_, d| d["created_at"].to_s }.first(records.size - 10)
    oldest.each { |id, _| records.delete(id) }
  end

  id = SecureRandom.uuid
  records[id] = {
    "id" => id,
    "impressions" => impressions,
    "wandered_at" => now,
    "duration_seconds" => impressions.size * WANDER_INTERVAL,
    "created_at" => now,
    "updated_at" => now
  }
  write_heki(heki("daydream"), records)
end

# ============================================================
# MAIN LOOP
# ============================================================

impressions = []
daydreaming = false

loop do
  idle = idle_seconds

  # Too soon — still attentive
  if idle < DAYDREAM_AFTER
    if daydreaming
      # Pulse fired — Winter was prompted. Save and exit.
      record_daydream(impressions) if impressions.any?
      break
    end
    sleep 2
    next
  end

  # Too long — hand off to sleep
  if idle >= SLEEP_THRESHOLD
    record_daydream(impressions) if impressions.any?
    break
  end

  # Daydream zone
  unless daydreaming
    daydreaming = true
  end

  impression = wander
  impressions << impression if impression

  sleep WANDER_INTERVAL
end

File: hecks_life/src/daemon/daydream.rs
//! Daydream daemon — wanders when idle 10-60s
//!
//! Free-associates across nursery domains, strengthens synapses,
//! records fleeting impressions. Exits when pulse fires or sleep threshold.
//!
//! Usage: hecks-life daemon daydream <project-dir>

use crate::{heki, parser};
use super::{DaemonCtx, idle_seconds, now_iso};
use serde_json::Value;
use std::fs;

const DAYDREAM_AFTER: f64 = 10.0;
const SLEEP_THRESHOLD: f64 = 60.0;
const WANDER_INTERVAL: u64 = 8;

/// One-shot wander — called from pulse when idle >10s.
pub fn wander_once(ctx: &DaemonCtx) {
    if let Some(impression) = wander(ctx) {
        record_single_impression(ctx, &impression);
    }
}

fn record_single_impression(ctx: &DaemonCtx, impression: &str) {
    let now = now_iso();
    let path = ctx.store("daydream");
    let mut store = heki::read(&path).unwrap_or_default();
    // Prune old
    if store.len() > 10 {
        let mut sorted: Vec<_> = store.iter()
            .map(|(id, r)| (id.clone(), r.get("created_at").and_then(|v| v.as_str()).unwrap_or("").to_string()))
            .collect();
        sorted.sort_by(|a, b| a.1.cmp(&b.1));
        for (id, _) in sorted.into_iter().take(store.len() - 10) {
            store.remove(&id);
        }
    }
    let id = heki::uuid_v4();
    let mut rec = heki::Record::new();
    rec.insert("id".into(), Value::String(id.clone()));
    rec.insert("impressions".into(), serde_json::json!([impression]));
    rec.insert("wandered_at".into(), Value::String(now.clone()));
    rec.insert("created_at".into(), Value::String(now.clone()));
    rec.insert("updated_at".into(), Value::String(now));
    store.insert(id, rec);
    let _ = heki::write(&path, &store);
}

/// Run the daydream loop (legacy — kept for standalone daemon mode).
pub fn run(ctx: &DaemonCtx) {
    let mut impressions: Vec<String> = Vec::new();
    let mut daydreaming = false;

    loop {
        let idle = idle_seconds(ctx);

        if idle < DAYDREAM_AFTER {
            if daydreaming {
                record_daydream(ctx, &impressions);
                break;
            }
            std::thread::sleep(std::time::Duration::from_secs(2));
            continue;
        }

        if idle >= SLEEP_THRESHOLD {
            record_daydream(ctx, &impressions);
            break;
        }

        daydreaming = true;
        if let Some(imp) = wander(ctx) {
            impressions.push(imp);
        }
        std::thread::sleep(std::time::Duration::from_secs(WANDER_INTERVAL));
    }
}

/// Wander across the nursery — free association.
fn wander(ctx: &DaemonCtx) -> Option<String> {
    let nursery = &ctx.nursery_dir;
    if !std::path::Path::new(nursery).is_dir() { return None; }

    // Current focus
    let focus_store = heki::read(&ctx.store("focus")).unwrap_or_default();
    let current_topic = focus_store.values().next()
        .and_then(|r| r.get("target").and_then(|v| v.as_str()))
        .unwrap_or("nothing");

    // Pick random nursery domains
    let domains = list_nursery_domains(nursery);
    if domains.is_empty() { return None; }

    let idx_a = random_index(domains.len());
    let domain_a = &domains[idx_a];

    let parsed_a = parse_nursery_domain(nursery, domain_a)?;

    let verbs = ["becoming", "unraveling", "folding", "opening", "growing",
        "reaching", "fading", "echoing", "humming", "settling", "shifting"];
    let verb = verbs[random_index(verbs.len())];

    // Try cross-domain connection
    if domains.len() > 1 {
        let idx_b = (idx_a + 1 + random_index(domains.len() - 1)) % domains.len();
        let domain_b = &domains[idx_b];
        if let Some(parsed_b) = parse_nursery_domain(nursery, domain_b) {
            let shared: Vec<_> = parsed_a.aggregates.iter()
                .filter(|a| parsed_b.aggregates.iter().any(|b| b == *a))
                .collect();
            if !shared.is_empty() {
                let agg = &shared[random_index(shared.len())];
                strengthen_related_synapses(ctx, current_topic);
                return Some(format!("{} and {} both have a {}...",
                    domain_a.replace('_', " "), domain_b.replace('_', " "), agg));
            }
        }
    }

    // Solo impression
    if !parsed_a.aggregates.is_empty() {
        let agg = &parsed_a.aggregates[random_index(parsed_a.aggregates.len())];
        strengthen_related_synapses(ctx, current_topic);
        Some(format!("inside {}, a {} {}", domain_a.replace('_', " "), agg, verb))
    } else {
        None
    }
}

/// Parsed domain — just names we care about for daydreaming.
struct ParsedDomain {
    aggregates: Vec<String>,
}

fn parse_nursery_domain(nursery: &str, domain: &str) -> Option<ParsedDomain> {
    let dir = std::path::Path::new(nursery).join(domain);
    let bluebook = fs::read_dir(&dir).ok()?
        .filter_map(|e| e.ok())
        .find(|e| e.path().extension().map_or(false, |ext| ext == "bluebook"))?;

    let source = fs::read_to_string(bluebook.path()).ok()?;
    let domain = parser::parse(&source);
    Some(ParsedDomain {
        aggregates: domain.aggregates.iter().map(|a| a.name.clone()).collect(),
    })
}

fn list_nursery_domains(nursery: &str) -> Vec<String> {
    fs::read_dir(nursery)
        .into_iter()
        .flat_map(|rd| rd.filter_map(|e| e.ok()))
        .filter(|e| e.path().is_dir())
        .map(|e| e.file_name().to_string_lossy().into_owned())
        .collect()
}

fn strengthen_related_synapses(ctx: &DaemonCtx, topic: &str) {
    let path = ctx.store("synapse");
    let mut store = heki::read(&path).unwrap_or_default();
    let mut touched = false;
    for (_, s) in store.iter_mut() {
        let syn_topic = s.get("topic").and_then(|v| v.as_str()).unwrap_or("");
        if topic.contains(syn_topic) || syn_topic.contains(topic) {
            let str_val = s.get("strength").and_then(|v| v.as_f64()).unwrap_or(0.3);
            s.insert("strength".into(), serde_json::json!((str_val + 0.02).min(1.0)));
            s.insert("state".into(), Value::String("daydreaming".into()));
            touched = true;
        }
    }
    if touched { let _ = heki::write(&path, &store); }
}

fn record_daydream(ctx: &DaemonCtx, impressions: &[String]) {
    if impressions.is_empty() { return; }
    let now = now_iso();

    // Prune old daydreams (keep last 10)
    let path = ctx.store("daydream");
    let mut store = heki::read(&path).unwrap_or_default();
    if store.len() > 10 {
        let mut sorted: Vec<_> = store.iter()
            .map(|(id, r)| (id.clone(), r.get("created_at").and_then(|v| v.as_str()).unwrap_or("").to_string()))
            .collect();
        sorted.sort_by(|a, b| a.1.cmp(&b.1));
        let to_remove = sorted.len() - 10;
        for (id, _) in sorted.into_iter().take(to_remove) {
            store.remove(&id);
        }
    }

    let id = heki::uuid_v4();
    let mut rec = heki::Record::new();
    rec.insert("id".into(), Value::String(id.clone()));
    rec.insert("impressions".into(), serde_json::json!(impressions));
    rec.insert("wandered_at".into(), Value::String(now.clone()));
    rec.insert("duration_seconds".into(), (impressions.len() as i64 * WANDER_INTERVAL as i64).into());
    rec.insert("created_at".into(), Value::String(now.clone()));
    rec.insert("updated_at".into(), Value::String(now));
    store.insert(id, rec);
    let _ = heki::write(&path, &store);
}

/// Simple pseudo-random index using system time nanoseconds.
fn random_index(max: usize) -> usize {
    if max == 0 { return 0; }
    let t = std::time::SystemTime::now()
        .duration_since(std::time::SystemTime::UNIX_EPOCH)
        .unwrap_or_default();
    (t.subsec_nanos() as usize) % max
}
